grobid-trainer/src/main/java/org/grobid/trainer/evaluation/EvaluationUtilities.java

```diff
@@ -1,371 +1,349 @@
-		theResult = theResult.replace("\n\n", "\n \n");
-		StringTokenizer stt = new StringTokenizer(theResult, "\n");
-		boolean allGood = true;
-		int correctInstance = 0;
-		int totalInstance = 0;
-		String line = null;
-		while (stt.hasMoreTokens()) {
-			line = stt.nextToken();
-			if ((line.trim().length() == 0) || (!stt.hasMoreTokens())) {
-				// instance done
-				totalInstance++;
-				if (allGood) {
-					correctInstance++;
-				}
-				// we reinit for a new instance
-				allGood = true;
-			} else {
-				StringTokenizer st = new StringTokenizer(line, "\t ");
-				String obtainedLabel = null;
-				String expectedLabel = null;
-				while (st.hasMoreTokens()) {
-					obtainedLabel = getPlainLabel(st.nextToken());
-					if (st.hasMoreTokens()) {
-						expectedLabel = obtainedLabel;
-					}
-				}
-
-				if (!obtainedLabel.equals(expectedLabel)) {
-					// one error is enough to have the whole instance false, damn!
-					allGood = false;
-				}
-			}
-		}
-
-		report.append(String.format("%-27s %d\n", "Total expected instances:", totalInstance));
-		report.append(String.format("%-27s %d\n", "Correct instances:", correctInstance));
-		double accuracy = (double) correctInstance / (totalInstance);
-		report.append(String.format("%-27s %s\n",
-				"Instance-level recall:",
-				TextUtilities.formatTwoDecimals(accuracy * 100)));
-	
-		return report.toString();
-	}
-
-	public static Stats tokenLevelStats(String theResult) {
-		Stats wordStats = new Stats();
-		String line = null;
-		StringTokenizer stt = new StringTokenizer(theResult, System.lineSeparator());
-		while (stt.hasMoreTokens()) {
-			line = stt.nextToken();
-
-			if (line.trim().length() == 0) {
-				continue;
-			}
-			// the two last tokens, separated by a tabulation, gives the
-			// expected label and, last, the resulting label -> for Wapiti
-			StringTokenizer st = new StringTokenizer(line, "\t ");
-			String obtainedLabel = null;
-			String expectedLabel = null;
-
-			while (st.hasMoreTokens()) {
-				obtainedLabel = getPlainLabel(st.nextToken());
-				if (st.hasMoreTokens()) {
-					expectedLabel = obtainedLabel;
-				}
-			}
-			
-			if ((expectedLabel == null) || (obtainedLabel == null)) {
-				continue;
-			}
-
-			processCounters(wordStats, obtainedLabel, expectedLabel);
+            throw new GrobidException("An exception occurred while evaluating Grobid.", e);
+        }
+
+        return computeStats(theResult);
+    }
+
+    public static ModelStats computeStats(String theResult) {
+        ModelStats modelStats = new ModelStats();
+        // report token-level results
+        Stats wordStats = tokenLevelStats(theResult);
+        modelStats.setTokenStats(wordStats);
+
+        // report field-level results
+        Stats fieldStats = fieldLevelStats(theResult);
+        modelStats.setFieldStats(fieldStats);
+
+        // instance-level: instances are separated by a new line in the result file
+        // third pass
+        theResult = theResult.replace("\n\n", "\n \n");
```

Comment:
I believe that is copied from the current code. I find it a bit hard to follow. This seem to replace two new line feeds with with two new line feeds but a space in-between them. I guess in general it would maybe be easier if this method didn't have to parse the result. But it's probably out of scope.